daemon/cluster/executor/container/adapter.go

```diff
@@ -530,3 +530,14 @@ func isUnknownContainer(err error) bool {
 func isStoppedContainer(err error) bool {
 	return strings.Contains(err.Error(), "is already stopped")
 }
+
+func isErrContainerUnhealthy(err error) bool {
+	if err != nil {
+		if e, ok := err.(*exitError); ok {
+			if c := e.Cause(); c != nil {
+				return strings.Contains(c.Error(), "unhealthy container")
+			}
```

Comment:
We should probably avoid using `errors.Cause()` to prevent issues if something lower in the stack starts using native "go" wrapping of errors.

In this case, I'm wondering if just checking `strings.Contains()` on the `exitError` would suffice.

That said, if I see correctly, these errors are generated in https://github.com/moby/moby/blob/07d60bc2571ba3d680f21adc84d87803ab4959c6/daemon/cluster/executor/container/controller.go#L248-L273

Which is in this codebase, so could we instead wrap a specific error that we can check for here? (so modify the code I linked above, and make it wrap a typed error, so that we can check for that error-type instead of do string-matching?

Reply:
Sorry, I don't quite understand what you mean. Here, i just want to filter out `exitError{code: ctnr.State.ExitCode, cause: healthErr} `. If I don't extract `cause` in `exitError`, I don't have a way to catch it.
Yes, It's better to check error-type instead of do string-matching for `healthErr`.  I will try doing it.

Reply:
Modified the code per your suggestion.